### ISSUE TITLE
How to create a customized model architecture that fits sentence-transformer's training framework?

### ISSUE THREAD

**Issue by HuangLED** [UKPLab/sentence-transformers#3390]
I'd like to train a two tower model that takes categorical features, floats features in one tower, and the other tower just encodes a document using an out of the box embedding.  Then the outputs from both towers are feed into sentence transformers loss function.  All the training configuration should reuse sentence transformer's setup (loss function implementation, Training Arguments, etc) as much as possible.  

Is this even feasible?  Skimmed through the document found this page here (https://www.sbert.net/docs/sentence_transformer/usage/custom_models.html#structure-of-sentence-transformer-models), but the example on this page seems to be creating a new module, but only as part of a purely sequential models, each connected to its next. 

Much appreciated!

**Comment by tomaarsen** [comment]
Hello!

You've got great timing - we're actively refactoring the Asym (asymmetric) module into a Router module in a major upcoming update ([pull request here for now](https://github.com/arthurbr11/sentence-transformers/pull/1)). Notably, this Router module is designed (but not exclusively) for separating query and document processing. This will be a common usage: 

![Image](https://github.com/user-attachments/assets/eb564040-5e00-4062-bb9f-fc4a9c66cf8a)

Here, the query side uses a [static embedding model](https://huggingface.co/blog/static-embeddings) whereas the document side uses a more standard transformers-based embedding model. You can train this like normal, with one addition: the new `router_mapping` argument that maps (training) dataset columns to routes so the model knows what tower to use. See also this snippet from our upcoming documentation:

![Image](https://github.com/user-attachments/assets/78a0ffc9-1943-426e-8da9-21ce19e9abaf)

We'll be launching with the v5.0 release quite soon, so I would recommend holding tight until then. But if you really can't wait, you can start with [Asym](https://sbert.net/docs/package_reference/sentence_transformer/models.html#sentence_transformers.models.Asym) - we'll just be changing that one a bit (e.g. no more training and inferencing with singleton dictionaries).

- Tom Aarsen

**Comment by HuangLED** [comment]
This is awesome Tom! Excited! 

definitely will hold for it being released?   what is our ETA? :-D 


Also a follow up question, since the static embedding model takes float inputs, then my input definitely needs to define a customized Dataloader I guess.  If so, what is the recommended practice between:1) stick with the new Trainer & TrainingArgument interface?  Or 2)  the legacy model.fit(train_objectives=, ...) API ?  
   
I traced down both options, seems like Option#1 requires some override and hack into your nice class def.  Option#2 is cleaner, but I am guess it is called legacy for a strong reason (no?).
